homeassistant/components/wake/__init__.py

```diff
@@ -0,0 +1,165 @@
+    """Unload a config entry."""
+    component: EntityComponent = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+class WakeWordDetectionEntity(RestoreEntity):
+    """Represent a single WWD provider."""
+
+    _attr_should_poll = False
+
+    @property
+    @abstractmethod
+    def supported_wake_words(self) -> list[WakeWord]:
+        """Return a list of supported wake words."""
+
+    @abstractmethod
+    async def async_process_audio_stream(
+        self, stream: AsyncIterable[tuple[bytes, int]]
+    ) -> DetectionResult | None:
+        """Try to detect wake word(s) in an audio stream with timestamps.
+
+        Audio must be 16Khz sample rate with 16-bit mono PCM samples.
+        """
+
+
+@websocket_api.websocket_command(
+    vol.All(
+        websocket_api.BASE_COMMAND_MESSAGE_SCHEMA.extend(
+            {
+                vol.Required("type"): "wake/detect",
+                vol.Optional("entity_id"): vol.Any(str, None),
+                vol.Optional("timestamp_start"): vol.Any(int, None),
+            },
+        ),
+    ),
+)
+@websocket_api.async_response
+async def websocket_detect(
+    hass: HomeAssistant,
+    connection: websocket_api.ActiveConnection,
+    msg: dict[str, Any],
+) -> None:
+    """Try to detect wake word(s) from a binary audio stream.
+
+    Audio must be 16Khz sample rate with 16-bit mono PCM samples.
+    """
+    timestamp_ms = msg.get("timestamp_start", 0)
+    entity_id = msg.get("entity_id", async_default_engine(hass))
+    engine = async_get_wake_word_detection_entity(hass, entity_id)
+
+    if engine is None:
+        raise ValueError("Wake word provider not found")
+
+    handler_id: int | None = None
+    unregister_handler: Callable[[], None] | None = None
+
+    # Audio pipeline that will receive audio as binary websocket messages
+    audio_queue: asyncio.Queue[bytes] = asyncio.Queue()
+
+    async def wwd_stream() -> AsyncGenerator[tuple[bytes, int], None]:
+        nonlocal timestamp_ms
+
+        # Yield until we receive an empty chunk
+        while chunk := await audio_queue.get():
+            chunk_ms = (len(chunk) / 2) // 16  # 16-bit mono @16KHz
+            timestamp_ms += chunk_ms
+            yield (chunk, timestamp_ms)
+
+    def handle_binary(
+        _hass: HomeAssistant,
+        _connection: websocket_api.ActiveConnection,
+        data: bytes,
+    ) -> None:
+        # Forward to WWD audio stream
+        audio_queue.put_nowait(data)
+
+    handler_id, unregister_handler = connection.async_register_binary_handler(
+        handle_binary
+    )
+
+    # Confirm subscription
+    connection.send_result(msg["id"], {"handler_id": handler_id})
+
+    run_task = hass.async_create_task(engine.async_process_audio_stream(wwd_stream()))
+
+    # Cancel pipeline if user unsubscribes
+    connection.subscriptions[msg["id"]] = run_task.cancel
+
+    try:
+        result = await run_task
+        connection.send_message({} if result is None else asdict(result))
```

Comment:
This is wrong. `send_message` can only be used with either result or error messages, and cannot be used when in a subscription (like this command). All messages need to be send using `send_event`